### PR TITLE
[MLLIB]: Avoid divided by zero norm, cache normalized Word Vectors to speed up

### PR DESCRIPTION
**Pull request by ezli**
1. Check if norm == 0 when calculate the division, add a ScalaTest for divided by zero scenario;
2. Cache the normalized wordVectors, speed up multiple findSynonyms() calls; 
3. Do lazy loading for wordVectors and wordVectorsNormalized; 
4. Normalize fVector in findSynonyms() to make cosine distances comparable across all words.

### CODE REVIEW

**Comment by srowen** [comment]
Have a look at https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark first please. You need a JIRA. I'd also break this down at least into logical commits to review.

**Comment by ezli** [comment]
Thanks for you comment. I will do that.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala [hunk 1]
```diff
@@ -540,6 +536,21 @@ class Word2VecModel private[mllib] (
...
+   * @param vector An array to be normalized
+   * @return a new normalized array
+   */
+  def euclideanNormalize(vector: Array[Float]):Array[Float] = {
+    val norm = blas.snrm2(vector.size, vector, 1)
+
+    if (norm == 0) {
+      Array.fill[Float](vector.size)(0)
+    } else {
+      vector.map(v => v/norm)
```
Comment (kashif) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: perhaps you could use `blas.scal` instead?

**Comment by srowen** [comment]
@ezli for the several new JIRAs you've opened, add `[SPARK-xxxx]` to the title if you intend to keep this open. It might be better to close this PR and make a new one with the changes for each JIRA in separate commits, at least. You can also open multiple PRs; use your judgment about how interrelated they are.

**Comment by ezli** [comment]
Close this PR and create a new one corresponding to the Jira ticket.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala (hunk 1 above)
Comment (ezli) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: Good point. I have incorporated this in the new PR: https://github.com/apache/spark/pull/6245

**Pull request closed without merging**